File: benchmarks/beginner/B3-array-sum/csharp/Program.cs
using System;
using System.Diagnostics;

const int N = 10_000_000;
const int RUNS = 10;
const long EXPECTED = 50000005000000;
double[] timings = new double[RUNS];

Console.WriteLine("=== BENCHMARK: B3 - Array Sum ===");
Console.WriteLine("Language: C#");
Console.WriteLine($"Runtime:  .NET {Environment.Version}");
Console.WriteLine();

long result = 0;
Stopwatch sw = new();

for (int i = 0; i < RUNS; i++)
{
    sw.Restart();
    long[] arr = new long[N];
    for (int j = 0; j < N; j++)
    {
        arr[j] = j + 1;
    }
    long s = 0;
    foreach (long v in arr)
    {
        s += v;
    }
    result = s;
    sw.Stop();

    timings[i] = sw.Elapsed.TotalMilliseconds;
    Console.WriteLine($"Run {i + 1,2}: {timings[i]:F2} ms");
}

Console.WriteLine();
double min = timings[0], max = timings[0], sum = 0;
foreach (double t in timings)
{
    sum += t;
    if (t < min) 
    {
        min = t;
    }

    if (t > max)
    {
       max = t; 
    }
}
double avg = sum / RUNS;
double variance = 0;
foreach (double t in timings)
{
    variance += (t - avg) * (t - avg);
}
variance /= RUNS;
double stddev = Math.Sqrt(variance);

Console.WriteLine($"Min:    {min:F2} ms");
Console.WriteLine($"Avg:    {avg:F2} ms");
Console.WriteLine($"Max:    {max:F2} ms");
Console.WriteLine($"StdDev: {stddev:F2} ms");
Console.WriteLine();
string status = result == EXPECTED ? "✓" : "✗";
Console.WriteLine($"Verification: {result} {status}");

File: benchmarks/beginner/B3-array-sum/fsharp/Program.fs
open System
open System.Diagnostics

[<EntryPoint>]
let main (argv: string array) =
    let n = 10_000_000
    let runs = 10
    let expected = 50000005000000L
    let timings = Array.zeroCreate<float> runs

    printfn "=== BENCHMARK: B3 - Array Sum ==="
    printfn "Language: F#"
    printfn "Runtime:  .NET %O" Environment.Version
    printfn ""

    let mutable result = 0L
    let sw = Stopwatch()

    for i in 0 .. runs - 1 do
        sw.Restart()
        let arr = Array.init n (fun j -> int64 (j + 1))
        let mutable s = 0L
        for j in 0 .. n - 1 do
            s <- s + arr.[j]
        result <- s
        sw.Stop()

        timings.[i] <- sw.Elapsed.TotalMilliseconds
        printfn "Run %2d: %.2f ms" (i + 1) timings.[i]

    printfn ""
    let minT = Array.min timings
    let maxT = Array.max timings
    let avg = Array.average timings
    let variance = timings |> Array.averageBy (fun t -> (t - avg) ** 2.0)
    let stddev = sqrt variance

    printfn "Min:    %.2f ms" minT
    printfn "Avg:    %.2f ms" avg
    printfn "Max:    %.2f ms" maxT
    printfn "StdDev: %.2f ms" stddev
    printfn ""
    let status = if result = expected then "✓" else "✗"
    printfn "Verification: %d %s" result status
    0
